DOCS-440 create general robot configuration doc v2 (#890)
viamrobotics/docs: docs/manage/configuration.md

---
title: "Robot Configuration"
linkTitle: "Robot Configuration"
weight: 25
type: "docs"

> description: "Details about configuring robots with Viam."

Author: Done % processe*s*

tags: ["manage", "components"]
---

> To be able to use a robot, you must first configure it.

Reviewer: I've tried an alternative flow for the config file and the config tab sections:

- it starts by explaining how a configuration is used and how it works
- it sees the app-based flow as the default and explains that
- I think the way to include that you can also store the configuration on the robot itself is as a note that points to a section at the end of the page (see below). So users use the app to create the configuration and if they cannot have the robot connect to the internet, store it directly on the robot. That's the best user experience I think. Slack me if you want to discuss that more. 

```
A robot's configuration tells the code running the robot what *resources*, [hardware *components*](#components) and [software *services*](#services), it has access to, as well as any relevant parameters for those resources.

To configure your robot, go to the [Viam app](https://app.viam.com) and [create a new robot](http://localhost:1313/installation/install/) and follow the steps on the **SETUP** page.

The setup steps copy your robots credentials to your robot and store them at <file>/etc/viam.json</file>.
The credentials look like this:

```json
{"cloud":{"app_address":"https://app.viam.com:123","id":"123456ab-a1b2-12a3-a123-1a2b345cdefg6",
"secret":"ab0cd0efghijkl0m12no3pqrstuvw0xyza4bcd567891abcd"}}
```

When you turn on your robot, `viam-server` starts up and checks its local <file>/etc/viam.json</file> for its configuration information.
The robot uses these credentials to get its full config from the Viam app.
To begin with, that configuration is nearly empty.

When you completed the setup steps and successfully connected to your robot, go to the **CONFIG** tab.

{{< alert title="Note" color="note" >}}
Your robot does not need to stay connected to the Viam app after it has obtained its configuration file.
The configuration is cached locally.
If you cannot connect your robot to the internet at all, you can also [store the configuration file](#Local Setup) on the robot itself.
{{< /alert >}}

## The CONFIG tab

The **CONFIG** tab is the place to configure everything about your robot.

![The CONFIG tab of the Viam app with a box highlighting the CONFIG tab and the Mode selector (where you can select Builder or Raw JSON).](../img/configuration/config-tab.png)

You can use the mode selector to switch between **Builder** and **Raw JSON**:

- **Builder** mode provides a graphical interface for configuring your robot resources.
- **Raw JSON** mode provides a text editing field where you can write and edit the config manually.

Regardless of the mode you choose, Viam stores the configuration file in [JSON (JavaScript Object Notation)](https://en.wikipedia.org/wiki/JSON).

If you add components in **Builder** mode and click **Save Config** at the bottom of the screen, you can switch to **Raw JSON** and see the corresponding JSON generated by the builder.

{{% expand "An example config file for a robot with a board component, motor component, and vision service configured" %}}

```json
{
  "components": [
    {
      "name": "local",
      "type": "board",
      "model": "pi",
      "attributes": {},
      "depends_on": []
    },
    {
      "name": "my-motor",
      "type": "motor",
      "model": "gpio",
      "attributes": {
        "pins": {
          "a": "13",
          "b": "15"
        },
        "board": "local",
        "max_rpm": 120
      },
      "depends_on": [],
    }
  ],
  "services": [
    {
      "name": "detector",
      "type": "vision",
      "attributes": {
        "register_models": [
          {
            "parameters": {
              "segment_size_px": 200,
              "hue_tolerance_pct": 0.05,
              "detect_color": "#19FFD9"
            },
            "type": "color_detector",
            "name": "green_detector"
          }
        ]
      }
    }
  ],
  "modules": []
}
```

{{% /expand %}}

Your robot's **CONFIG** page has sub-tabs to configure:

- [components](#components): Components are the hardware of your robot.
- [services](#services): Services are the software that runs on your robot.
- [modules](#modules): Modular resources are a way to add resource types or models that are not built into Viam.
- [remotes](#remotes): Remote are a way to connect two separate robots so one can access the resources of the other.
- [processes](#processes): Processes automatically run specified scripts when the robot boots.
- [fragments](#fragments): Fragments are a way of sharing and managing identical configuration files (or parts of config files) across multiple robots.
- [frames](#frame-system): Frames hold reference frame information for the relative position of components in space.
<!--[authorization keys](#authnetwork) -->
```

You can use the mode selector to switch between **Builder** and **Raw JSON**:

- **Builder** mode provides a graphical interface for configuring your robot resources.
It generates properly formatted JSON for you.
- **Raw JSON** mode provides a text editing field where you can write and edit the full config manually.
It can be convenient to view a large config in raw JSON mode because all the components, services, modules and more that you configure in individual tabs of builder mode will appear in one long file in raw JSON mode.
You can also copy and paste the raw JSON if needed.

![The CONFIG tab of the Viam app with a box highlighting the CONFIG tab and the Mode selector (where you can select Builder or Raw JSON).](../img/configuration/config-tab.png)

If you add components in **Builder** mode and click **Save Config** at the bottom of the screen, you can switch to **Raw JSON** and see the JSON that has been generated by the builder.

The sections of the robot config (which each have corresponding sub-tabs on the **CONFIG** tab) are as follows:

## Components

Components represent the pieces of hardware on your robot that you want to control with Viam.

> Find information on how to configure each supported component type in its respective [documentation](/components/).

Author: I'm trying to let them know that they can just jump to the specifics and not bother with this page

> Find information on how to configure each supported component type in its respective [documentation](/components/).

Reviewer: How would you feel about moving this second sentence to before "When you configure a component on the **CONFIG** tab, it"?

The problem with it is that you are essentially devaluing this piece of content. You're before even giving them the general information saying - just go to this other page. And I think this content is good. So either we can make the link to the components less like they should just immediately go there or move it below.

You can choose any unique name for a component.

- `type`: The broad component category, such as `motor`, `arm` or `camera`.
  Components of a given type have a common API.

- `model`: Indicates the more specific category of hardware.
Components of the same model are supported using the same low-level code.

- `attributes`: A struct to define things like how the component is wired to the robot, its dimensions, and other specifications; attributes vary widely between models.
  See the [component documentation](/components/) for a given component type and model for more details.

- `depends_on`: Any components that a given component relies upon, and that must be initialized on boot before this component is initialized.
  Many built-in components have convenient implicit dependencies, in which case `depends_on` can be left blank.
  For example, a [`gpio` motor](/components/motor/gpio/) depends on the `board` to which it is wired, but it has a dedicated `board` attribute and `viam-server` will automatically initialize that board before it looks for the motor.

> When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.

Author: Ooh I would but i just moved a different sentence to above this and want to call *it* out as a tip too so then it will look weird. hmm

> When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.

Reviewer: maybe put the enitre segment into it
```suggestion
{{% alert title="Tip" color="tip" %}}
Find specific information on how to configure each supported component type in its respective [documentation](/components/).

When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.
Meanwhile the **CODE SAMPLE** tab will also update to include code for some basic interaction with that component using the Viam [SDKs](/program/sdk-as-client/).
{{% /alert %}}
```

> When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.

Author: Mmm but those are two very different concepts. Maybe we should just have two back to back alerts? Would that be weird?

> When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.

Reviewer: Yeah. I think it's fine to have with the space in between

> When you configure a component on the **CONFIG** tab, it will also appear on the **CONTROL** tab which gives you an interface to test and interact with it.

Author: Done!

The `name` serves as an identifier when accessing the resource from your code, as well as when configuring other resources that are dependent on that resource.
You can choose any unique name for a service.

The other aspects of configuring a service are highly specific to the type of service.
See the [services documentation](/services/) for more information.

## Modules

[Modular resources](/program/extend/modular-resources/) are a way to add resource types or models that are not built into Viam.

To add a modular resource as a component or service of your robot, configure a module per the [modular resource documentation](/program/extend/modular-resources/#add-a-module-to-your-robot-configuration).

## Remotes

Configuring a remote is a way to connect two separate robots so one can access the resources of the other.
Find more information in our remotes documentation.

## Processes

To automatically run a specified script when the robot boots, configure a *process*.

Start by giving the process a **Name** (`id` in raw JSON) (an identifier of your choice) and clicking **Create Process**.

> The following fields will expand (raw JSON attribute names shown in code style in parentheses):
> 
> - **Executable** (`name`): (string) The command you want to execute when your robot connects to the server.
> - **Arguments** (`args`): (string) Optional arguments to follow the command
> - **Working Directory** (`cwd`): (string) Where you want the process to execute.

Author: ```suggestion
Then fill in the following fields:

- **Executable** (`name`): The command you want to execute when your robot connects to the server.
- **Arguments** (`args`): Optional arguments to follow the command.
- **Working Directory** (`cwd`): Where you want the process to execute.
```

To add a fragment to a robot:

- Go to the **FRAGMENTS** sub-tab of your robot's **CONFIG** tab on the [Viam app](https://app.viam.com).
- Look through the list of available fragments and click **ADD** next to any fragments you want to add to your robot.
- Click **Save Config** at the bottom of the screen.

{{% alert title="Note" color="note" %}}

The components or other resources included in the fragment will *not* appear in the **COMPONENTS** sub-tab or in the `components` section of your config.
You will simply see the fragment ID in the `fragments` section of your config, but all the resources *will* appear on the **CONTROL** tab when your robot comes on line, and they will be accessible from SDK code as regular resources of your robot.

{{% /alert %}}

To create a fragment, go to [https://app.viam.com/fragments](https://app.viam.com/fragments).

For an example of adding a fragment to a robot, see the [Viam Rover fragment tutorial](/try-viam/rover-resources/rover-tutorial-fragments/).
<!--
## Auth/network
-->

## Frame system

> 

Author: Git being weird but as you can see I did apply this suggestion!

  It can take up to around 20 seconds for components to initialize on the **CONTROL** tab.

## Troubleshooting

If you run into issues, here are some things to try:

- Check the **LOGS** tab to check for any errors or other info from `viam-server`.
- You can also view logs locally by running the following command on the robot:
  
  {{< tabs >}}
  {{% tab name="Linux" %}}

  ```bash
  sudo journalctl --unit=viam-server
  ```

  {{% /tab %}}
  {{% tab name="macOS" %}}

  ```bash
  cat $(brew --prefix)/var/log/viam.log
  ```

  {{% /tab %}}
  {{< /tabs >}}

- Make sure all configured components are actually saved to your config.
   If they aren't, you will see an **Unsaved Changes** note next to the **Save Config** button at the bottom of the config screen.
- Try restarting `viam-server` by clicking **Restart** in the upper right corner of the app.
   It takes a few minutes for the server to shut down and restart.
- Make sure the issue is not hardware related.
  Some things to check are that the robot has adequate power, all wires are properly connected, and no chips or other hardware components are shorted or overheated.

> - {{< snippet "social.md" >}}

Author: and I applied this one!

---
title: "Configure a Robot"
linkTitle: "Configure a Robot"
weight: 25
type: "docs"
description: "Before you can program a robot, you must configure its components and services as well as any modules, remotes, processes and frames."
tags: ["manage", "components"]
---

Before you can program a robot, you must configure it.

A robot's configuration tells the code running the robot what *resources* (hardware *components* and software *services*) it has access to, as well as any relevant parameters for those resources.

> To configure your robot, go to the [Viam app](https://app.viam.com), [create a new robot](/installation/install/) and follow the steps on the **SETUP** page.

Reviewer: ```suggestion
To configure your robot, go to the [Viam app](https://app.viam.com), [create a new robot](/installation/) and follow the steps on the **SETUP** page.

```

Author: This was the link you suggested in the alternate flow thing so I applied it, but I’m actually wondering if it doesn’t make sense because “create a robot” =/= install. Maybe this sentence should actually be 
```suggestion
To start configuring, go to the [Viam app](https://app.viam.com), create a new robot and follow the steps on your new robot’s **SETUP** tab.
```

Reviewer: Sure

## Services

[Services](/services/) are built-in software packages that make it easier to add complex capabilities such as motion planning or object detection to your robot.

For services, the `type` specifies which of the Viam services you want to use on your robot, such as the Vision Service or the Motion Service.

The `name` serves as an identifier when accessing the resource from your code, as well as when configuring other resources that are dependent on that resource.
You can choose any unique name for a service.

The other aspects of configuring a service are highly specific to the type of service.
See the [services documentation](/services/) for more information.

## Modules

[Modular resources](/program/extend/modular-resources/) are a way to add resource types or models that are not built into Viam.

To add a modular resource as a component or service of your robot, configure a module per the [modular resource documentation](/program/extend/modular-resources/#add-a-module-to-your-robot-configuration).

## Remotes

Configuring a remote is a way to connect two separate robots so one can access the resources of the other.

> Find more information in our remotes documentation.

Reviewer: link missing

Author: ```suggestion
Find more information in our [remotes documentation](../../manage/parts-and-remotes).
```

Find more information in the [frame system documentation](/services/frame-system/).

## Troubleshooting

If you run into issues, here are some things to try:

- Check the **LOGS** tab to check for any errors or other info from `viam-server`.
- You can also view logs locally by running the following command on the robot:
  
  {{< tabs >}}
  {{% tab name="Linux" %}}

  ```bash
  sudo journalctl --unit=viam-server
  ```

  {{% /tab %}}
  {{% tab name="macOS" %}}

  ```bash
  cat $(brew --prefix)/var/log/viam.log
  ```

  {{% /tab %}}
  {{< /tabs >}}

- Make sure all configured components are actually saved to your config.
   If they aren't, you will see an **Unsaved Changes** note next to the **Save Config** button at the bottom of the config screen.
- Try restarting `viam-server` by clicking **Restart** in the upper right corner of the app.
   It takes a few minutes for the server to shut down and restart.
- Make sure the issue is not hardware related.
  Some things to check are that the robot has adequate power, all wires are properly connected, and no chips or other hardware components are shorted or overheated.
- {{< snippet "social.md" >}}

> ## Local Setup

Reviewer: Do you think there's a good place above where we could link to this?

Author: We do link to it from the first section
